internal/parser/translate_httproute_test.go

```diff
@@ -0,0 +1,445 @@
+			routes: []*gatewayv1alpha2.HTTPRoute{{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "basic-httproute",
+					Namespace: corev1.NamespaceDefault,
+				},
+				Spec: gatewayv1alpha2.HTTPRouteSpec{
+					CommonRouteSpec: gatewayv1alpha2.CommonRouteSpec{
+						ParentRefs: []gatewayv1alpha2.ParentRef{{
+							Name: gatewayv1alpha2.ObjectName("fake-gateway"),
+						}},
+					},
+					Rules: []gatewayv1alpha2.HTTPRouteRule{{
+						Matches: []gatewayv1alpha2.HTTPRouteMatch{
+							{
+								Path: &gatewayv1alpha2.HTTPPathMatch{
+									Type:  &pathMatchRegex,
+									Value: kong.String("httpbin$"),
+								},
+							},
+							{
+								Path: &gatewayv1alpha2.HTTPPathMatch{
+									Type:  &pathMatchPrefix,
+									Value: kong.String("/httpbin"),
+								},
+							},
+						},
+						BackendRefs: []gatewayv1alpha2.HTTPBackendRef{{
+							BackendRef: gatewayv1alpha2.BackendRef{
+								BackendObjectReference: gatewayv1alpha2.BackendObjectReference{
+									Name: gatewayv1alpha2.ObjectName("fake-service"),
+									Port: &httpPort,
+								},
+							},
+						}},
+					}},
+				},
+			}},
+			expected: ingressRules{
+				SecretNameToSNIs: SecretNameToSNIs{},
+				ServiceNameToServices: map[string]kongstate.Service{
+					"default.fake-service.80": {
+						Service: kong.Service{ // only 1 service should be created
+							ConnectTimeout: kong.Int(60000),
+							Host:           kong.String("fake-service.default.80.svc"),
+							Name:           kong.String("default.fake-service.80"),
+							Path:           kong.String("/"),
+							Port:           kong.Int(80),
+							Protocol:       kong.String("http"),
+							ReadTimeout:    kong.Int(60000),
+							Retries:        kong.Int(5),
+							WriteTimeout:   kong.Int(60000),
+						},
+						Backend: kongstate.ServiceBackend{
+							Name: "fake-service",
+							Port: kongstate.PortDef{
+								Mode:   kongstate.PortMode(1),
+								Number: 80,
+							},
+						},
+						Namespace: "default",
+						Routes: []kongstate.Route{{ // only 1 route should be created
+							Route: kong.Route{
+								Name: kong.String("httproute.default.basic-httproute.1"),
+								Paths: []*string{
+									kong.String("/httpbin"),
+								},
+								PreserveHost: kong.Bool(true),
+								Protocols: []*string{
+									kong.String("http"),
+									kong.String("https"),
+								},
+								StripPath: kong.Bool(true),
+							},
+							Ingress: util.K8sObjectInfo{
+								Name:        "basic-httproute",
+								Namespace:   corev1.NamespaceDefault,
+								Annotations: make(map[string]string),
+							},
+						}},
+						K8sService: corev1.Service{},
+					},
+				},
+			},
+			errMsgs: []string{
+				"regular expression path matches are not yet supported",
+			},
+		},
+	} {
+		// generate a logger so we can capture logged errors
+		logger, output := newTestLoggerForTranslators()
```

Comment:
asserting on log prints is going to be a maintenance challenge and looks like a code smell to me - because it's not testing by public interface.

Suggestion: change the interface of `fromHTTPRoutes` to return a collection of errors.

This will also be a step in the direction of error reporting by means of status fields.

Reply:
To me this would start to creep into making changes to the machinery above this translator, which I feel **strongly** we should avoid as a part of this PR and follow up on holistically in https://github.com/Kong/kubernetes-ingress-controller/issues/2130 rather than making this PR responsible for also taking on general improvements to how these translators work.

Reply:
Adding a test that makes an implicit dependency on log structure (as opposed to using the general/public interface of the function) is a software quality issue to me, and I'm not comfortable with this decision. We're in charge of the interface of this function; if it doesn't fit our needs, let's update it.

Can you think of some other way to ensure that the interface is sufficient for testing?

Reply:
I can think of other ways but they're all even more scope creep. I've opted to just go with your original suggestion here for speed. 98bb709eabd030a4e5d18fad16fdd59bc2dc4f5c